File: v2v_cooperative_pkg/v2v_cooperative_pkg/safety_supervisor_node.py
#!/usr/bin/env python3
"""
safety_supervisor_node
───────────────────────
전체 예외 상황 감시 및 안전 정지 노드 (Safety Layer).

어떤 상태에서든 아래 조건 중 하나가 충족되면 /safe_stop = True 발행:

우선순위 (높은 순):
  P1. LiDAR 근접 장애물 (scan 기반)
  P2. 인식 불안정 (vision/lane 토픽 타임아웃)
  P3. 경로 오류 (REVERSE_EXECUTE 중 /reverse_path_ready = False 지속)
  P4. 협상 불일치 (양쪽 모두 '진행' 결정)
  P5. 제어 이상 (cmd_vel 이상값)

/safe_stop 해제는 모든 조건이 해소되면 자동으로 이루어진다.

입력 토픽:
  /scan                    sensor_msgs/LaserScan
  /relative_distance       Float32
  /reverse_path_ready      Bool
  /vehicle_state           String
  /decision_result         String
  /cmd_vel                 geometry_msgs/Twist
  /lane_status_active      String

출력 토픽:
  /safe_stop               Bool
  /supervisor_status       String  (디버깅용)
"""

import math
from typing import Optional

import rclpy
from geometry_msgs.msg import Twist
from rclpy.node import Node
from sensor_msgs.msg import LaserScan
from std_msgs.msg import Bool, Float32, String


class SafetySupervisorNode(Node):
    def __init__(self) -> None:
        super().__init__('safety_supervisor_node')

        # ── 파라미터 ─────────────────────────────────────────────────────
        self.declare_parameter('supervisor_hz', 20.0)

        # P1: LiDAR 근접 장애물
        self.declare_parameter('lidar_min_dist_m', 0.15)          # 전방 안전 거리
        self.declare_parameter('lidar_front_angle_deg', 60.0)     # 전방 감시 각도 범위 ±
        self.declare_parameter('lidar_rear_angle_deg', 30.0)      # 후방 감시 각도 범위 ± (후진 중)

        # P2: 인식 타임아웃
        self.declare_parameter('vision_timeout_sec', 1.0)
        self.declare_parameter('lane_timeout_sec', 1.0)

        # P3: 경로 오류 지속 시간
        self.declare_parameter('path_error_hold_sec', 2.0)

        # P5: 제어 이상
        self.declare_parameter('max_safe_linear', 0.20)
        self.declare_parameter('max_safe_angular', 2.0)

        self.supervisor_hz       = float(self.get_parameter('supervisor_hz').value)
        self.lidar_min_dist_m    = float(self.get_parameter('lidar_min_dist_m').value)
        self.lidar_front_angle   = math.radians(float(self.get_parameter('lidar_front_angle_deg').value))
        self.lidar_rear_angle    = math.radians(float(self.get_parameter('lidar_rear_angle_deg').value))
        self.vision_timeout_sec  = float(self.get_parameter('vision_timeout_sec').value)
        self.lane_timeout_sec    = float(self.get_parameter('lane_timeout_sec').value)
        self.path_error_hold_sec = float(self.get_parameter('path_error_hold_sec').value)
        self.max_safe_linear     = float(self.get_parameter('max_safe_linear').value)
        self.max_safe_angular    = float(self.get_parameter('max_safe_angular').value)

        # ── 내부 상태 ─────────────────────────────────────────────────────
        self.scan_msg:          Optional[LaserScan] = None
        self.relative_distance: Optional[float]     = None
        self.reverse_path_ready: bool = False
        self.vehicle_state:     str = 'NORMAL_CENTER_DRIVE'
        self.decision_result:   str = 'none'
        self.cmd_vel:           Optional[Twist] = None
        self.lane_status:       str = 'unknown'

        self.last_scan_stamp:     Optional[float] = None
        self.last_dist_stamp:     Optional[float] = None
        self.last_lane_stamp:     Optional[float] = None
        self.path_error_start:    Optional[float] = None

        self.safe_stop: bool = False
        self.stop_reason: str = ''

        # ── 구독 ─────────────────────────────────────────────────────────
        self.create_subscription(LaserScan, '/scan',               self._cb_scan,    10)
        self.create_subscription(Float32,   '/relative_distance',  self._cb_dist,    10)
        self.create_subscription(Bool,      '/reverse_path_ready', self._cb_path,    10)
        self.create_subscription(String,    '/vehicle_state',      self._cb_state,   10)
        self.create_subscription(String,    '/decision_result',    self._cb_decision, 10)
        self.create_subscription(Twist,     '/cmd_vel',            self._cb_cmd,     10)
        self.create_subscription(String,    '/lane_status_active', self._cb_lane,    10)

        # ── 발행 ─────────────────────────────────────────────────────────
        self.pub_stop   = self.create_publisher(Bool,   '/safe_stop',          10)
        self.pub_status = self.create_publisher(String, '/supervisor_status',  10)

        dt = 1.0 / self.supervisor_hz if self.supervisor_hz > 0 else 0.05
        self.create_timer(dt, self._step)
        self.get_logger().info('safety_supervisor_node 시작')

    # ── 콜백 ─────────────────────────────────────────────────────────────
    def _cb_scan(self, msg: LaserScan) -> None:
        self.scan_msg = msg
        self.last_scan_stamp = self._now()

    def _cb_dist(self, msg: Float32) -> None:
        self.relative_distance = float(msg.data)
        self.last_dist_stamp = self._now()

    def _cb_path(self, msg: Bool) -> None:
        self.reverse_path_ready = bool(msg.data)

    def _cb_state(self, msg: String) -> None:
        self.vehicle_state = str(msg.data).strip()

    def _cb_decision(self, msg: String) -> None:
        self.decision_result = str(msg.data).strip().lower()

    def _cb_cmd(self, msg: Twist) -> None:
        self.cmd_vel = msg

    def _cb_lane(self, msg: String) -> None:
        self.lane_status = str(msg.data).strip().lower()
        self.last_lane_stamp = self._now()

    # ── 유틸 ─────────────────────────────────────────────────────────────
    def _now(self) -> float:
        return self.get_clock().now().nanoseconds * 1e-9

    def _fresh(self, stamp: Optional[float], timeout: float) -> bool:
        if stamp is None:
            return False
        return (self._now() - stamp) <= timeout

    # ── P1: LiDAR 근접 장애물 ─────────────────────────────────────────────
    def _check_lidar(self) -> Optional[str]:
        if self.scan_msg is None or not self._fresh(self.last_scan_stamp, 0.3):
            return None  # 스캔 없으면 이 조건은 무시

        scan = self.scan_msg
        is_reversing = self.vehicle_state == 'REVERSE_EXECUTE'
        # 감시 각도 범위 선택
        if is_reversing:
            watch_angle = self.lidar_rear_angle
            # 후방 기준: π ± rear_angle
            center_angle = math.pi
        else:
            watch_angle = self.lidar_front_angle
            center_angle = 0.0

        min_range = float('inf')
        n = len(scan.ranges)
        for i, r in enumerate(scan.ranges):
            if not math.isfinite(r) or r <= 0.0:
                continue
            angle = scan.angle_min + i * scan.angle_increment
            # 각도 정규화 [-π, π]
            diff = math.atan2(math.sin(angle - center_angle), math.cos(angle - center_angle))
            if abs(diff) <= watch_angle:
                min_range = min(min_range, r)

        if min_range < self.lidar_min_dist_m:
            return f'P1_lidar_close min_range={min_range:.2f}m'
        return None

    # ── P2: 인식 타임아웃 ─────────────────────────────────────────────────
    def _check_perception_timeout(self) -> Optional[str]:
        # 주행 중(정지 상태 아닐 때)에만 체크
        if self.vehicle_state in ('SAFE_STOP', 'WAIT_PASS'):
            return None
        if not self._fresh(self.last_lane_stamp, self.lane_timeout_sec):
            return 'P2_lane_timeout'
        return None

    # ── P3: 경로 오류 ─────────────────────────────────────────────────────
    def _check_path_error(self) -> Optional[str]:
        if self.vehicle_state != 'REVERSE_EXECUTE':
            self.path_error_start = None
            return None
        if self.reverse_path_ready:
            self.path_error_start = None
            return None
        if self.path_error_start is None:
            self.path_error_start = self._now()
            return None
        if (self._now() - self.path_error_start) >= self.path_error_hold_sec:
            return 'P3_no_reverse_path'
        return None

    # ── P4: 협상 불일치 ───────────────────────────────────────────────────
    def _check_negotiation_conflict(self) -> Optional[str]:
        # 양쪽 모두 '진행'이면 충돌 (간이 감지: 상대 led_state로 판단)
        # 여기서는 decision_result='proceed' && vehicle_state가 진행 중인데
        # 상대도 접근 중인 상황을 체크
        # (정밀 검증은 V2V 통신 구현 후 확장)
        return None  # 기본 비활성 (추후 확장)

    # ── P5: 제어 이상 ─────────────────────────────────────────────────────
    def _check_control_anomaly(self) -> Optional[str]:
        if self.cmd_vel is None:
            return None
        lx = abs(float(self.cmd_vel.linear.x))
        az = abs(float(self.cmd_vel.angular.z))
        if lx > self.max_safe_linear or az > self.max_safe_angular:
            return f'P5_control_anomaly linear={lx:.2f} angular={az:.2f}'
        return None

    # ── 메인 스텝 ─────────────────────────────────────────────────────────
    def _step(self) -> None:
        reasons = []

        r = self._check_lidar()
        if r:
            reasons.append(r)

        r = self._check_perception_timeout()
        if r:
            reasons.append(r)

        r = self._check_path_error()
        if r:
            reasons.append(r)

        r = self._check_negotiation_conflict()
        if r:
            reasons.append(r)

        r = self._check_control_anomaly()
        if r:
            reasons.append(r)

        if reasons:
            if not self.safe_stop:
                self.stop_reason = ' | '.join(reasons)
                self.get_logger().warn(f'[Safety] SAFE_STOP: {self.stop_reason}')
            self.safe_stop = True
        else:
            if self.safe_stop:
                self.get_logger().info('[Safety] 안전 조건 해소 → safe_stop 해제')
            self.safe_stop = False
            self.stop_reason = 'ok'

        # 발행
        stop_msg = Bool(); stop_msg.data = self.safe_stop
        self.pub_stop.publish(stop_msg)

        status = self.stop_reason if self.safe_stop else 'ok'
        s = String(); s.data = status
        self.pub_status.publish(s)


def main(args=None) -> None:
    rclpy.init(args=args)
    node = SafetySupervisorNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == '__main__':
    main()

File: v2v_cpp_nodes/src/safety_supervisor_node.cpp
/**
 * safety_supervisor_node (C++)
 * ─────────────────────────────
 * 20 Hz 전체 예외 상황 감시 노드.
 *
 * 우선순위:
 *   P1. LiDAR 근접 장애물
 *   P2. 인식 타임아웃
 *   P3. 경로 오류 (REVERSE_EXECUTE 중 path 미수신)
 *   P4. 제어 이상 (cmd_vel 이상값)
 */

#include <cmath>
#include <string>
#include <vector>
#include <algorithm>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/bool.hpp"
#include "std_msgs/msg/float32.hpp"
#include "std_msgs/msg/string.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "sensor_msgs/msg/laser_scan.hpp"

class SafetySupervisorNode : public rclcpp::Node {
public:
    SafetySupervisorNode() : Node("safety_supervisor_node") {
        // ── 파라미터 ─────────────────────────────────────────────
        this->declare_parameter("supervisor_hz",         20.0);
        this->declare_parameter("lidar_min_dist_m",      0.15);
        this->declare_parameter("lidar_front_angle_deg", 60.0);
        this->declare_parameter("lidar_rear_angle_deg",  30.0);
        this->declare_parameter("lane_timeout_sec",      1.0);
        this->declare_parameter("path_error_hold_sec",   2.0);
        this->declare_parameter("max_safe_linear",       0.20);
        this->declare_parameter("max_safe_angular",      2.0);
        this->declare_parameter("startup_grace_sec",     3.0);

        supervisor_hz_       = this->get_parameter("supervisor_hz").as_double();
        lidar_min_dist_m_    = this->get_parameter("lidar_min_dist_m").as_double();
        lidar_front_angle_   = this->get_parameter("lidar_front_angle_deg").as_double() * M_PI / 180.0;
        lidar_rear_angle_    = this->get_parameter("lidar_rear_angle_deg").as_double() * M_PI / 180.0;
        lane_timeout_sec_    = this->get_parameter("lane_timeout_sec").as_double();
        path_error_hold_sec_ = this->get_parameter("path_error_hold_sec").as_double();
        max_safe_linear_     = this->get_parameter("max_safe_linear").as_double();
        max_safe_angular_    = this->get_parameter("max_safe_angular").as_double();
        startup_grace_sec_   = this->get_parameter("startup_grace_sec").as_double();
        startup_time_        = now_sec();

        // ── 구독 ────────────────────────────────────────────────
        sub_scan_  = create_subscription<sensor_msgs::msg::LaserScan>(
            "/scan", 10,
            [this](sensor_msgs::msg::LaserScan::SharedPtr m) {
                scan_ = m; last_scan_stamp_ = now_sec();
            });

        sub_state_ = create_subscription<std_msgs::msg::String>(
            "/vehicle_state", 10,
            [this](std_msgs::msg::String::SharedPtr m) { vehicle_state_ = m->data; });

        sub_path_rdy_ = create_subscription<std_msgs::msg::Bool>(
            "/reverse_path_ready", 10,
            [this](std_msgs::msg::Bool::SharedPtr m) { reverse_path_ready_ = m->data; });

        sub_cmd_ = create_subscription<geometry_msgs::msg::Twist>(
            "/cmd_vel", 10,
            [this](geometry_msgs::msg::Twist::SharedPtr m) { cmd_vel_ = *m; });

        sub_lane_ = create_subscription<std_msgs::msg::String>(
            "/lane_status_active", 10,
            [this](std_msgs::msg::String::SharedPtr m) {
                lane_status_ = m->data; last_lane_stamp_ = now_sec();
            });

        // ── 발행 ────────────────────────────────────────────────
        pub_stop_   = create_publisher<std_msgs::msg::Bool>("/safe_stop", 10);
        pub_status_ = create_publisher<std_msgs::msg::String>("/supervisor_status", 10);

        double dt = (supervisor_hz_ > 0) ? (1.0 / supervisor_hz_) : 0.05;
        timer_ = create_wall_timer(
            std::chrono::duration<double>(dt),
            std::bind(&SafetySupervisorNode::step, this));

        RCLCPP_INFO(get_logger(), "safety_supervisor_node (C++) 시작");
    }

private:
    double now_sec() const {
        return get_clock()->now().nanoseconds() * 1e-9;
    }
    bool in_startup_grace() const {
        return (now_sec() - startup_time_) < startup_grace_sec_;
    }
    bool is_fresh(double stamp, double timeout) const {
        return (stamp > 0) && (now_sec() - stamp <= timeout);
    }

    // ── P1: LiDAR 근접 ────────────────────────────────────────────
    std::string check_lidar() {
        if (!scan_ || !is_fresh(last_scan_stamp_, 0.5))
            return "";

        bool reversing = (vehicle_state_ == "REVERSE_EXECUTE");
        double center_angle = reversing ? M_PI : 0.0;
        double watch = reversing ? lidar_rear_angle_ : lidar_front_angle_;

        double min_r = 1e9;
        for (size_t i = 0; i < scan_->ranges.size(); ++i) {
            float r = scan_->ranges[i];
            if (!std::isfinite(r) || r <= 0.0f) continue;
            double angle = scan_->angle_min
                         + static_cast<double>(i) * scan_->angle_increment;
            double diff = std::atan2(std::sin(angle - center_angle),
                                     std::cos(angle - center_angle));
            if (std::abs(diff) <= watch)
                min_r = std::min(min_r, static_cast<double>(r));
        }
        if (min_r < lidar_min_dist_m_)
            return "P1_lidar_close=" + std::to_string(min_r);
        return "";
    }

    // ── P2: 인식 타임아웃 ───────────────────────────────────────────
    std::string check_perception() {
        if (vehicle_state_ == "SAFE_STOP" || vehicle_state_ == "WAIT_PASS")
            return "";
        if (in_startup_grace()) return "";
        if (!is_fresh(last_lane_stamp_, lane_timeout_sec_))
            return "P2_lane_timeout";
        return "";
    }

    // ── P3: 경로 오류 ──────────────────────────────────────────────
    std::string check_path_error() {
        if (vehicle_state_ != "REVERSE_EXECUTE") {
            path_error_start_ = 0.0;
            return "";
        }
        if (reverse_path_ready_) {
            path_error_start_ = 0.0;
            return "";
        }
        if (path_error_start_ <= 0.0)
            path_error_start_ = now_sec();
        if (now_sec() - path_error_start_ >= path_error_hold_sec_)
            return "P3_no_reverse_path";
        return "";
    }

    // ── P4: 제어 이상 ──────────────────────────────────────────────
    std::string check_control() {
        if (std::abs(cmd_vel_.linear.x) > max_safe_linear_ ||
            std::abs(cmd_vel_.angular.z) > max_safe_angular_)
            return "P4_control_anomaly";
        return "";
    }

    // ── 메인 루프 ───────────────────────────────────────────────────
    void step() {
        std::vector<std::string> reasons;
        auto push = [&](const std::string &s) {
            if (!s.empty()) reasons.push_back(s);
        };
        push(check_lidar());
        push(check_perception());
        push(check_path_error());
        push(check_control());

        bool stop = !reasons.empty();

        if (stop && !safe_stop_) {
            std::string all;
            for (auto &r : reasons) all += r + " | ";
            RCLCPP_WARN(get_logger(), "[Safety] SAFE_STOP: %s", all.c_str());
        }
        if (!stop && safe_stop_) {
            RCLCPP_INFO(get_logger(), "[Safety] 해제");
        }
        safe_stop_ = stop;

        std_msgs::msg::Bool b;
        b.data = safe_stop_;
        pub_stop_->publish(b);

        std_msgs::msg::String s;
        if (reasons.empty()) {
            s.data = "ok";
        } else {
            s.data = reasons.front();
        }
        pub_status_->publish(s);
    }

    // ── 파라미터 ─────────────────────────────────────────────────
    double supervisor_hz_, lidar_min_dist_m_;
    double lidar_front_angle_, lidar_rear_angle_;
    double lane_timeout_sec_, path_error_hold_sec_;
    double max_safe_linear_, max_safe_angular_;
    double startup_grace_sec_, startup_time_;

    // ── 상태 ─────────────────────────────────────────────────────
    sensor_msgs::msg::LaserScan::SharedPtr scan_;
    double last_scan_stamp_{0.0}, last_lane_stamp_{0.0};
    std::string vehicle_state_{"NORMAL_CENTER_DRIVE"};
    std::string lane_status_{"unknown"};
    bool reverse_path_ready_{false};
    geometry_msgs::msg::Twist cmd_vel_;
    bool safe_stop_{false};
    double path_error_start_{0.0};

    // ── ROS 인터페이스 ────────────────────────────────────────────
    rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr sub_scan_;
    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr       sub_state_, sub_lane_;
    rclcpp::Subscription<std_msgs::msg::Bool>::SharedPtr         sub_path_rdy_;
    rclcpp::Subscription<geometry_msgs::msg::Twist>::SharedPtr   sub_cmd_;
    rclcpp::Publisher<std_msgs::msg::Bool>::SharedPtr            pub_stop_;
    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr          pub_status_;
    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<SafetySupervisorNode>());
    rclcpp::shutdown();
    return 0;
}
